### PR TITLE
DSPDC-508 Initialize ftp-lib

### PR DESCRIPTION
**Pull request by danxmoran**
With functionality to:
* Read objects
* List directories

### CODE REVIEW

File: ftp/src/main/scala/org/broadinstitute/monster/storage/ftp/FtpEventLogger.scala [hunk 1]
```diff
@@ -0,0 +1,31 @@
...
+
+    // Make sure we don't log sensitive info
+    val sanitized = if ("PASS".equalsIgnoreCase(cmd)) {
+      "PASS ******"
+    } else {
+      event.getMessage
+    }
+
+    logger
+      .debug(s"Sent FTP command with message: $sanitized")
```
Comment (benjamincarlin) [inline ftp/src/main/scala/org/broadinstitute/monster/storage/ftp/FtpEventLogger.scala:23]: Would it be helpful to show the command?

**Review by benjamincarlin — APPROVED** [review]

File: ftp/src/it/scala/org/broadinstitute/monster/storage/ftp/FtpApiIntegrationSpec.scala [hunk 2]
```diff
@@ -0,0 +1,262 @@
...
+import org.scalatest.{EitherValues, FlatSpec, Matchers}
+
+import scala.concurrent.ExecutionContext
+
+class FtpApiIntegrationSpec extends FlatSpec with Matchers with EitherValues {
+
+  private implicit val cs: ContextShift[IO] = IO.contextShift(ExecutionContext.global)
+
+  // FIXME: Change this to point to our own FTP server once we have one up and running.
+  // NCBI's site is maddeningly slow, you probably want to take a break once you launch these tests...
```
Comment (raaidbroad) [inline ftp/src/it/scala/org/broadinstitute/monster/storage/ftp/FtpApiIntegrationSpec.scala:15]: lol

File: ftp/src/it/scala/org/broadinstitute/monster/storage/ftp/FtpApiIntegrationSpec.scala [hunk 3]
```diff
@@ -0,0 +1,262 @@
...
+    "release_notes" -> FtpApi.Directory,
+    "submission_templates" -> FtpApi.Directory,
+    "tab_delimited" -> FtpApi.Directory,
+    "vcf_GRCh37" -> FtpApi.Directory,
+    "vcf_GRCh38" -> FtpApi.Directory,
+    "xml" -> FtpApi.Directory,
+    "xsd_public" -> FtpApi.Directory,
+    "xsd_submission" -> FtpApi.Directory
+  )
+  private val testEmptyDir = "pub/README"
```
Comment (raaidbroad) [inline ftp/src/it/scala/org/broadinstitute/monster/storage/ftp/FtpApiIntegrationSpec.scala]: do we have any test file conventions of putting private vals up top? If so, maybe move these up! If not, nevermind 😇

File: ftp/src/it/scala/org/broadinstitute/monster/storage/ftp/FtpApiIntegrationSpec.scala [hunk 4]
```diff
@@ -0,0 +1,262 @@
...
+      .flatMap(_.listDirectory(testEmptyDir))
+      .compile
+      .toList
+      .unsafeRunSync()
+
+    contents shouldBe Nil
+  }
+
+  it should "raise a helpful error if listing a nonexistent directory" in {
+    val fakePath = "foobar"
```
Comment (raaidbroad) [inline ftp/src/it/scala/org/broadinstitute/monster/storage/ftp/FtpApiIntegrationSpec.scala]: nitpick: change to fakeDir for consistency with previous tests? Not super important though

File: ftp/src/main/scala/org/broadinstitute/monster/storage/ftp/FtpApi.scala [hunk 5]
```diff
@@ -0,0 +1,303 @@
...
+        .scan(start -> Either.right[Throwable, Chunk[Byte]](Chunk.empty)) {
+          case ((n, _), Right(bytes)) => (n + bytes.size, Right(bytes))
+          case ((n, _), Left(err))    => (n, Left(err))
+        }
+        .flatMap {
+          case (_, Right(bytes)) =>
+            Stream.chunk(bytes)
+          case (n, Left(err: FTPConnectionClosedException)) =>
+            val message = s"Hit connection error while reading $path, retrying"
+            Stream.eval(logger.warn(err)(message)) >> readStream(n)
```
Comment (raaidbroad) [inline ftp/src/main/scala/org/broadinstitute/monster/storage/ftp/FtpApi.scala]: should there be any wait time or max number of retries for the recursive call?

File: ftp/src/main/scala/org/broadinstitute/monster/storage/ftp/FtpApi.scala [hunk 6]
```diff
@@ -0,0 +1,303 @@
...
+            .flatMap(_ => client.listRemoteDirectory(path))
+
+          Stream.eval(listDir).flatMap(Stream.emits).map { ftpFile =>
+            val tpe = ftpFile.getType match {
+              case FTPFile.FILE_TYPE          => RegularFile
+              case FTPFile.DIRECTORY_TYPE     => Directory
+              case FTPFile.SYMBOLIC_LINK_TYPE => Symlink
+              case _                          => Other
+            }
+            (ftpFile.getName, tpe)
```
Comment (raaidbroad) [inline ftp/src/main/scala/org/broadinstitute/monster/storage/ftp/FtpApi.scala]: Is `tpe` a typo? Did you mean `type` or is this an abbreviation that's going over my head 😅

File: ftp/src/main/scala/org/broadinstitute/monster/storage/ftp/FtpApi.scala [hunk 7]
```diff
@@ -0,0 +1,303 @@
...
+    /** Use FTP to list the contents of a remote directory. */
+    def listRemoteDirectory(path: String): IO[List[FTPFile]]
+  }
+
+  /**
+    * Enum representation of FTP file types.
+    *
+    * TODO: This representation isn't bound to FTP. It'd probably be worth breaking
+    * out a new 'common' project to hold its definition, and then have each lib adapt
+    * their specific methods to use it.
```
Comment (raaidbroad) [inline ftp/src/main/scala/org/broadinstitute/monster/storage/ftp/FtpApi.scala:176]: 👍

File: ftp/src/main/scala/org/broadinstitute/monster/storage/ftp/FtpApi.scala (hunk 6 above)
Comment (danxmoran) [reply ftp/src/main/scala/org/broadinstitute/monster/storage/ftp/FtpApi.scala]: `type` is a keyword in Scala for defining type aliases. There isn't a clear "best practice" for the abbreviation to use when you want to use `type` to name a variable, but people usually go with either `tpe` or `typ`. I'm open to `typ` if it's more clear.

Comment (raaidbroad) [reply ftp/src/main/scala/org/broadinstitute/monster/storage/ftp/FtpApi.scala]: Got it. I think `ftpType` would be clear but ye `typ` works for me, `tpe` threw me haha

**Review by raaidbroad — APPROVED** [review]
🎉

**Pull request merged**